File: Backend/src/middleware/rate_limiter.rs
use axum::{
    extract::{ConnectInfo, Request},
    http::StatusCode,
    response::{IntoResponse, Response},
    Json,
};
use serde_json::json;
use std::collections::HashMap;
use std::net::SocketAddr;
use std::sync::Arc;
use tokio::sync::Mutex;

#[derive(Clone)]
pub struct RateLimiter {
    window_secs: u64,
    max_requests: u32,
    counters: Arc<Mutex<HashMap<String, (u32, std::time::Instant)>>>,
}

impl RateLimiter {
    pub fn new(window_secs: u64, max_requests: u32) -> Self {
        Self {
            window_secs,
            max_requests,
            counters: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Create a general API rate limiter (500 req/min per IP)
    pub fn general() -> Self {
        Self::new(60, 500)
    }

    /// Create an auth rate limiter (5 req/min per IP)
    pub fn auth() -> Self {
        Self::new(60, 5)
    }

    /// Create an AI rate limiter (20 req/min per IP)
    pub fn ai() -> Self {
        Self::new(60, 20)
    }

    /// Create an admin rate limiter (10000 req/min per IP)
    pub fn admin() -> Self {
        Self::new(60, 10000)
    }

    /// Extract client IP from request headers or fallback to socket info
    pub fn extract_client_ip(request: &Request) -> String {
        request
            .headers()
            .get("x-forwarded-for")
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.split(',').next())
            .map(|s| s.trim().to_string())
            .or_else(|| {
                request
                    .headers()
                    .get("x-real-ip")
                    .and_then(|v| v.to_str().ok())
                    .map(|s| s.to_string())
            })
            .or_else(|| {
                request
                    .extensions()
                    .get::<ConnectInfo<SocketAddr>>()
                    .map(|c| c.0.ip().to_string())
            })
            .unwrap_or_else(|| "unknown".to_string())
    }

    /// Check and increment rate limit counter. Returns Ok(()) if allowed, Err(Response) if limited.
    pub async fn check(&self, client_ip: &str) -> Result<(), Response> {
        let now = std::time::Instant::now();
        let mut counters = self.counters.lock().await;

        // Prune stale entries every ~10k entries to prevent unbounded growth
        if counters.len() > 10_000 {
            counters.retain(|_, (_, ts)| now.duration_since(*ts).as_secs() <= self.window_secs);
        }

        let entry = counters.entry(client_ip.to_string()).or_insert((0, now));

        if now.duration_since(entry.1).as_secs() > self.window_secs {
            *entry = (1, now);
        } else if entry.0 >= self.max_requests {
            return Err((
                StatusCode::TOO_MANY_REQUESTS,
                Json(json!({
                    "success": false,
                    "error_code": "RATE_LIMITED",
                    "message": "Too many requests. Please try again later.",
                })),
            )
                .into_response());
        } else {
            entry.0 += 1;
        }

        drop(counters);
        Ok(())
    }
}

File: Backend/python/app/middleware/rate_limiter.py
import time
from typing import Dict, Tuple
from fastapi import Request, HTTPException, status

class RateLimiter:
    def __init__(self, window_secs: int, max_requests: int):
        self.window_secs = window_secs
        self.max_requests = max_requests
        # Key: client_ip -> Value: (request_count, window_start_time)
        self.counters: Dict[str, Tuple[int, float]] = {}

    @classmethod
    def general(cls) -> "RateLimiter":
        return cls(60, 500)

    @classmethod
    def auth(cls) -> "RateLimiter":
        return cls(60, 5)

    @classmethod
    def ai(cls) -> "RateLimiter":
        return cls(60, 100)

    @classmethod
    def admin(cls) -> "RateLimiter":
        return cls(60, 10000)

    @staticmethod
    def extract_client_ip(request: Request) -> str:
        """Extract client IP address from headers (like Cloudflare, Nginx reverse proxy) or fallback."""
        x_forwarded_for = request.headers.get("x-forwarded-for")
        if x_forwarded_for:
            return x_forwarded_for.split(",")[0].strip()
        
        x_real_ip = request.headers.get("x-real-ip")
        if x_real_ip:
            return x_real_ip.strip()
            
        client = request.client
        if client:
            return client.host
            
        return "unknown"

    def check(self, client_ip: str) -> None:
        """
        Verify request count for the client IP.
        Raises 429 Too Many Requests if the limit is exceeded.
        """
        now = time.time()
        
        # Cleanup stale keys to avoid unbounded memory growth
        if len(self.counters) > 10000:
            self.counters = {
                ip: val for ip, val in self.counters.items()
                if now - val[1] <= self.window_secs
            }

        if client_ip not in self.counters:
            self.counters[client_ip] = (1, now)
            return

        count, start_time = self.counters[client_ip]

        if now - start_time > self.window_secs:
            # New window
            self.counters[client_ip] = (1, now)
        elif count >= self.max_requests:
            raise HTTPException(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                detail={
                    "success": False,
                    "error_code": "RATE_LIMITED",
                    "message": "Too many requests. Please try again later."
                }
            )
        else:
            self.counters[client_ip] = (count + 1, start_time)
